src/Compilers/CSharp/Portable/Binder/Binder_Patterns.cs

```diff
@@ -485,22 +484,68 @@ TypeSymbol BindRecursivePatternType(TypeSyntax typeSyntax, TypeSymbol inputType,
             }
         }
 
+        // Work around https://github.com/dotnet/roslyn/issues/20648: The compiler's internal APIs such as `declType.IsTupleType`
+        // do not correctly treat the non-generic struct `System.ValueTuple` as a tuple type.  We explicitly perform the tests
+        // required to identify it.  When that bug is fixed we should be able to remove this code and its callers.
+        internal static bool IsZeroElementTupleType(TypeSymbol type)
```

Comment:
>IsZeroElementTupleType [](start = 29, length = 22)

Would it be simpler go get the well-known type `ValueTuple` and compare them?

Reply:
No; there can be more than one tuple type imported from different assemblies.

---
In reply to: [234822985](https://github.com/dotnet/roslyn/pull/31056#discussion_r234822985) [](ancestors = 234822985)

Reply:
Sorry, one more question on this: would `IsTupleCompatible` do the job?

---
In reply to: [237273866](https://github.com/dotnet/roslyn/pull/31056#discussion_r237273866) [](ancestors = 237273866,234822985)